tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
+	idents, err := a.IdentFromFile(uri, filecontent)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ident := range idents {
+		if position.Line >= ident.StartLine && position.Line <= ident.EndLine {
+			return ident, nil
+		}
+	}
+
+	return nil, nil
+}
+
+
+// IdentFromFile gets all the Identifiers from a given BUILD file
+// *reads complete files only*
+func (a *Analyzer) IdentFromFile(uri lsp.DocumentURI, filecontent []string) ([]*Identifier,  error) {
+	filepath, err := GetPathFromURL(uri, "file")
+	if err != nil {
+		return nil, err
+	}
+	bytecontent, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	stmt, err := a.parser.ParseData(bytecontent, filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var idents []*Identifier
+	for i, val := range stmt {
+		//fmt.Println(val.Ident.Action.Call)
+		if val.Ident != nil {
+			ident := &Identifier{
+				IdentStatement: val.Ident,
+				// -1 from asp.Statement.Pos.Line, as lsp position requires zero index
+				StartLine: val.Pos.Line - 1,
+			}
+			idents = append(idents, ident)
+
+			// fillin End for the previous iterated identifier
+			if i > 0 {
+				fmt.Println(ident)
+				idents[i - 1].EndLine = ident.StartLine - 1
+			}
+		}
+	}
+
+	// Finally, we fill in the last of the endline
+	idents[len(idents) - 1].EndLine = len(filecontent)
+
+	return idents, nil
+}
+
+func newRuleDef(funcDef *asp.FuncDef) *RuleDef {
+	ruleDef := &RuleDef{
+		FuncDef:funcDef,
+		ArgMap: make(map[string]*Argument),
+	}
+
+	var headerSlice []string
+	header := "def " + funcDef.Name + "("
+	line := header
+	argLen := len(funcDef.Arguments)
+	if argLen > 0 {
+		for i, arg := range funcDef.Arguments {
+			// Check if it a builtin type method, and reconstruct header if it is
+			if i == 0 && arg.Name == "self" {
+				line = arg.Type[0] + "." +funcDef.Name + "("
+			} else {
+				// Fill in the ArgMap
+				argString := getArgument(arg)
+				required := true
+				if strings.Contains(argString, "=") {
+					required = false
+				}
+				ruleDef.ArgMap[arg.Name] = &Argument{
+					Argument: &arg,
+					definition: argString,
+					required:required,
+				}
+
+				// Add string to Header
+				if len(line) < 86 {
+					line += argString + ", "
+				} else {
+					headerSlice = append(headerSlice, line)
+					line = strings.Repeat(" ", len(header)) + argString + ", "
+				}
+			}
+
+			// finally, When we get to the end of the slice, append line
+			if i == argLen - 1 {
+				headerSlice = append(headerSlice, line)
+			}
+		}
+	}
+	joinedHeader := strings.Join(headerSlice, "\n")
+	ruleDef.Header = strings.TrimSuffix(joinedHeader, ", ") + ")"
+
+	return ruleDef
+}
+
+func getArgument(argument asp.Argument) string {
+	argType := strings.Join(argument.Type, "|")
+
+	// Get the default value for optional arguments
+	defaultVal := ""
+	if argument.Value != nil && argument.Value.Optimised != nil {
```

Comment:
Ah I see.

Probably not much we can do for now. Definitely think we should separate the builtin stuff - nothing outside `asp` should be using `Optimised` at all, we should probably embed the original files into the lsp and have the parser re-parse them (and not optimise).

That might make this reconstruction business easier too - you could just read the original lines in the file...